Use dprint's SQL plugin to format SQL files

thomaseizinger opened the pull request:
Thoughts? dprint all the things!

thomaseizinger committed:
Use dprint's SQL plugin to format SQL files

With the more resilient start-up behaviour, we can now reformat
our SQL files without the application crashing.

klochowicz reviewed (commented):
this solves one problem but creates another one - with this change we'd have a new db every release, even when there's no db changes. from user's pov it's almost the same an nuking the db. the fact that it's stored somewhere on disk or inside container it's hard to notice. 

Perhaps we could use a custom env variable that only gets bumped when we have a breaking db change?

thomaseizinger committed:
Add context inside function

If only one thing can fail inside the function, we can add the context
right there and avoid code duplication.

thomaseizinger committed:
Import anyhow::Result

thomaseizinger committed:
Encapsulate connecting to the database

thomaseizinger committed:
Backup DB and start empty if migrations fail

In case the migrations fail, we back up the database to a file including
the current timestamp and reconnect to an empty one.

This allows the application to startup and thus improves resilience.
It also allows us to make changes to our migration files.

Note on the `BoxFuture`: This is required to enable recursion of async
functions.

da-kami reviewed (approved):
Read through the decision making process. ack with creating a backup if the migration fails and then just start from scratch. We obviously have to be mindful about this in the future.

luckysori commented on daemon/src/db.rs:
```diff
@@ -11,17 +13,45 @@ use std::mem;
+            tracing::info!("Starting with a new database!");
+
+            // recurse to reconnect (async recursion requires a `BoxFuture`)
+            return connect(path).await;
```
🤔 So the assumption is that `run_migrations` cannot fail if the database has just been created, right? Otherwise I can see us creating an endless loop here.

thomaseizinger replied:
Yeah, these are the possible errors: https://docs.rs/sqlx-core/0.5.5/src/sqlx_core/migrate/error.rs.html#3

I guess there are some in here that are not resolved by using a new DB? We should probably only try again on `VersionMismatch` and `VersionMissing`. All others seem permanent!

luckysori commented on daemon/src/db.rs:
```diff
@@ -49,6 +49,8 @@ pub fn connect(path: PathBuf) -> BoxFuture<'static, Result<SqlitePool>> {
             return connect(path).await;
         }
 
+        tracing::info!("Opened database at {}", path.display());
```
🔧 Can probably squash this with the previous commit, unless there was a reason for it.

luckysori reviewed (approved):
Thank you for taking the time to figure this out and solve it cleanly. Inspiring PR :D

klochowicz commented on daemon/tests/harness/mod.rs:
```diff
@@ -365,17 +364,6 @@ macro_rules! deliver_event {
 }
 
-async fn in_memory_db() -> SqlitePool {
-    // Note: Every :memory: database is distinct from every other. So, opening two database
-    // connections each with the filename ":memory:" will create two independent in-memory
-    // databases. see: https://www.sqlite.org/inmemorydb.html
```
nit: any reason for getting rid of this comment? I for one found it useful (and it doesn't look like it can/will rot)

thomaseizinger replied:
No reason, just got lost in the refactor!

klochowicz reviewed (approved):
nice to see SQL files formatted 👍 approval pending the potential endless loop issue raised by @luckysori

luckysori reviewed (approved):
LGTM, thanks!
